services/QuillLMS/app/queries/quill_big_query.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+# All messages to BigQuery should use this module
+
+require 'google/api_client'
+require 'googleauth'
+require 'stringio'
+
+module QuillBigQuery
+  PROJECT_ID = 'analytics-data-stores'
+  FLOAT_FIELDS = %w(correct incorrect percentage)
+
+  class UnsupportedSchemaError < StandardError; end
+
+  def self.valid_schema?(json_body)
+    return false unless json_body['schema']['fields'].respond_to?(:count)
+    return false unless json_body['rows'].respond_to?(:count)
+
+    true
+  end
+
+  def self.floatify_fields(array_of_hashes)
+    array_of_hashes.map do |hsh|
+      hsh.each {|k,v| FLOAT_FIELDS.include?(k) && hsh[k] = v.to_f }
+    end
+    array_of_hashes
+  end
+
+  def self.transform_response(json_body)
+    raise UnsupportedSchemaError, json_body unless valid_schema?(json_body)
+
+    # Parsing - It returns row names and values separately in an odd setup
+    # This parses them down to [{"id" => 1, "name"=> "Dan", }, {"id" => 2, "name" => "Peter"}]
+    fields = json_body["schema"]["fields"].map {|h| h["name"]}
+    values = json_body["rows"].map {|r| r.values_at("f").flatten.map {|h| h.values.first}}
+    hash_results = values.map {|v| fields.zip(v).to_h}
+    floatify_fields(hash_results)
+  end
+
+  def self.get_response(query)
```

Comment:
Credit to Dan for writing the precursor to this method

Reply:
I would probably extract all of the client code (lines ~41-57) into its own file which allows for easier to stub out during testing while also providing a cleaner separation of concerns.

Reply:
I agree that there's a `BigQueryRequest` type of class here that should be extracted. It could be an `ApplicationService` with a main `run` method and broken down into several constants and submethods.